File: app/src/main/java/com/nimtego/plectrum/domain/interactor/InformationArtistInteractor.java
package com.nimtego.plectrum.domain.interactor;

import com.nimtego.plectrum.presentation.information_view.artist.model.ArtistDetailsModelK;
import com.nimtego.plectrum.presentation.main.model.AlbumModelK;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import dagger.internal.Preconditions;
import io.reactivex.Observable;

@Deprecated
public class InformationArtistInteractor
        extends BaseInteractor<ArtistDetailsModelK, InformationArtistInteractor.Params> {
    @Override
    protected Observable<ArtistDetailsModelK> buildUseCaseObservable(Params params) {
        Preconditions.checkNotNull(params);
        return repository.artistDetail(params.request).map(result -> {
            List<AlbumModelK> albumModels = duplicateRemove(Objects.requireNonNull(result.getAlbums()));
            //todo
            //result.setAlbums(albumModels);
            return result;
        });
    }

    private List<AlbumModelK> duplicateRemove(List<AlbumModelK> albumModels) {
        Map<String, AlbumModelK> duplicateCheck = new HashMap<>();
        albumModels
                .forEach(album -> duplicateCheck.put(album.getAlbumName(), album));
        return new ArrayList<>(duplicateCheck.values());
    }

    public static final class Params {

        private final String request;

        private Params(String request) {
            this.request = request;
        }

        public static Params forRequest(String request) {
            return new Params(request);
        }
    }
}

File: app/src/main/java/com/nimtego/plectrum/domain/interactor/InformationArtistInteractor.kt
package com.nimtego.plectrum.domain.interactor

import com.nimtego.plectrum.presentation.old.information_view.artist.model.ArtistDetailsModelK
import com.nimtego.plectrum.presentation.old.main.model.AlbumModel
import dagger.internal.Preconditions
import io.reactivex.Observable
import java.util.*

class InformationArtistInteractor : BaseInteractor<ArtistDetailsModelK, InformationArtistInteractor.Params>() {
    override fun buildUseCaseObservable(params: Params): Observable<ArtistDetailsModelK> {
        Preconditions.checkNotNull(params)
        return repository.artistDetail(params.request).map { result ->
            val albumModels = duplicateRemove(Objects.requireNonNull<List<AlbumModel>>(result.albums))
            //todo
            //result.setAlbums(albumModels);
            result
        }
    }

    private fun duplicateRemove(albumModels: List<AlbumModel>): List<AlbumModel> {
        val duplicateCheck = HashMap<String, AlbumModel>()
        albumModels.forEach {album ->
            album.albumName.let { duplicateCheck[it] = album }
        }
        return ArrayList(duplicateCheck.values)
    }

    class Params private constructor(val request: String) {
        companion object {
            fun forRequest(request: String): Params {
                return Params(request)
            }
        }
    }
}
